### DISCUSSION TITLE
Cuda agnostic compilation

### DISCUSSION THREAD

**Discussion by adrien-cotte** [openucx/ucx#6272] [Q&A] [answered]
Hello,

I have several Cuda versions (9.1, 10.0, 10.1, 10.2 and 11.0) and I would like to know if I must compile one UCX per Cuda or if I can have only one?

Needed area seems really strict:
```
$ ldd lib/ucx/libuct_cuda.so
libcudart.so.11.0 => not found 
```

Is it possible to link with ``libcudart.so`` or ``libcudart.so.11`` at least?

Thanks,
Adrien

**Comment by shamisp** [comment]
@bureddy maybe you can answer this ? thanks !

**Comment by bureddy** [comment, marked as answer]
@adrien-cotte different cuda runtimes are not compatible with each other.  things can go wrong by linking libcudart.so to a specific version. but you can have multiple cuda runtimes in the environment at the same time as long as the driver version is compatible (see below for compatibility matrix)
So, you can build one version of the UCX based on your driver version

```

CUDA Toolkit | Linux x86_64 Driver Version
-- | --
CUDA 11.1 (11.1.0) | >= 450.80.02
CUDA 11.0 (11.0.3) | >= 450.36.06
CUDA 10.2 (10.2.89) | >= 440.33
CUDA 10.1 (10.1.105) | >= 418.39
CUDA 10.0 (10.0.130) | >= 410.48
CUDA 9.2 (9.2.88) | >= 396.26
CUDA 9.1 (9.1.85) | >= 390.46
CUDA 9.0 (9.0.76) | >= 384.81
CUDA 8.0 (8.0.61 GA2) | >= 375.26
CUDA 8.0 (8.0.44) | >= 367.48
CUDA 7.5 (7.5.16) | >= 352.31
CUDA 7.0 (7.0.28) | >= 346.46

```
https://docs.nvidia.com/deploy/cuda-compatibility/index.html#binary-compatibility

**Reply by adrien-cotte** [reply]
Thank you for this answer!

**Reply by aruhela** [reply]
Adrien, did you figure out the configure arguments for this flexibility in cuda versions? Could you share how you build ucx?
